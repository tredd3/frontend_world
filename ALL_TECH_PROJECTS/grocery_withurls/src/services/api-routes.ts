const baseUrl = process.env.REACT_APP_BASE_URL || '';
const baseUrlLogin = process.env.REACT_APP_BASE_URL_LOGIN || '';

export const apiRoutes = {
  getNewAccessToken: `${baseUrl}/jm/auth/oauth/v2/token`,
  getDashboard: `${baseUrl}/jiokart/v1/jiogrocery/get/dashboard`,
  viewAllDashboard: `${baseUrl}/jiokart/v1/jiogrocery/view/all/dashboard`,
  getStores: `${baseUrl}/jiokart/v1/get/stores`,
  getCart: `${baseUrl}/jiokart/v1/jiogrocery/get/cart`,
  addCartItem: `${baseUrl}/jiokart/v1/add/cart/item`,
  removeCartItem: `${baseUrl}/jiokart/v1/remove/cart/item`,
  moveCartItem: `${baseUrl}/jiokart/v1/jiogrocery/move/item`,
  getCategories: `${baseUrl}/jiokart/v1/get/categories`,
  getProductList: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  checkServiceability: `${baseUrl}/jiokart/v1/jiogrocery/pincode/check`,
  getProductSearch: `${baseUrl}/jiokart/v1/jiogrocery/product/search`,
  getSuggestions: `${baseUrl}/jiokart/v1/jiogrocery/suggestion`,
  addUser: `${baseUrl}/jiokart/v1/jiogrocery/add/user`,
  saveAddress: `${baseUrl}/jiokart/v1/jiogrocery/save/user/address`,
  getAddress: `${baseUrl}/jiokart/v1/jiogrocery/get/user/address`,
  deleteAddress: `${baseUrl}/jiokart/v1/jiogrocery/remove/user/address`,
  getProductDetails: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  getSimilarItems: `${baseUrl}/jiokart/v1/jiogrocery/store/similar/items`,
  getOrderDetails: `${baseUrl}/jiokart/v1/jiogrocery/order/details`,
  getOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/listing`,
  getDeliveryModes: `${baseUrl}/jiokart/v1/get/delivery/modes`,
  getLocation: `${baseUrl}/jiokart/v1/jiogrocery/google/places`,
  searchOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/item/search`,
  checkout: `${baseUrl}/jiokart/v1/order/checkout`,
  placeOrder: `${baseUrl}/jiokart/v1/place/order`,
  getUserDetail: `${baseUrl}/jiokart/v1/jiogrocery/get/user`,
  getWishListData: `${baseUrl}/jiokart/v1/jiogrocery/get/item/list`,
  addItemToWishList: `${baseUrl}/jiokart/v1/jiogrocery/add/item`,
  removeItemFromWishList: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`,
  sendOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/send`,
  verifyOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/verify`,
  validateuser: `${baseUrl}/cr/v1/validateuser`,
  cancelOrder: `${baseUrl}/jiokart/v1/order/cancel`,
  config: `${baseUrl}/jiokart/v1/logged/customer/configurations`,
  updateUser: `${baseUrl}/jiokart/v1/jiogrocery/update/user`,
  orderRating: `${baseUrl}/jiokart/v1/order/rating`,
  getPaymentRequest: `${baseUrl}/jiokart/v1/get/payment/request`,
  getPaymentResponse: `${baseUrl}/jiokart/v1/get/payment/response`,
  getNotifications: `${baseUrl}/jiokart/v1/jiogrocery/notification/get`,
  saveNotification: `${baseUrl}/jiokart/v1/jiogrocery/notification/save`,
  clearNotifications: `${baseUrl}/jiokart/v1/jiogrocery/clear/notification`,
  removeCartSavedItem: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`
} as const;

const endpointsWithoutAuth = [
  apiRoutes.checkServiceability,
  apiRoutes.getStores,
  apiRoutes.getProductDetails,
  apiRoutes.getSimilarItems,
  apiRoutes.getLocation,
  apiRoutes.sendOtp,
  apiRoutes.getDashboard
];

const endpointsNeedingEnvelope = [
  apiRoutes.getCategories,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.getStores,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.getPaymentRequest,
  apiRoutes.getPaymentResponse,
  apiRoutes.orderRating
];

const endpointsNeedingStoreId = [
  apiRoutes.getCart,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.placeOrder,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.saveAddress,
  apiRoutes.getAddress,
  apiRoutes.getPaymentResponse,
  apiRoutes.removeCartSavedItem
];

export const needsEnvelope = (url: string) => endpointsNeedingEnvelope.includes(url);
export const needsAuthentication = (url: string) => !endpointsWithoutAuth.includes(url);
export const needsStoreId = (url: string) => endpointsNeedingStoreId.includes(url);

export const v2ApiRoutes = [
  apiRoutes.getPaymentResponse,
  apiRoutes.getOrderHistory,
  apiRoutes.getOrderDetails,
  apiRoutes.getCart,
  apiRoutes.getDeliveryModes
];
