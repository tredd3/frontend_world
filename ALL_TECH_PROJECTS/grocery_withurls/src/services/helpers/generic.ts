import { APIUser, UserPivot } from '../../types';

export const genderCodeToString = (code: APIUser['gender']): UserPivot['gender'] => {
  if (code === undefined) return undefined;
  return code === 1 ? 'male' : 'female';
};

export const genderStringToCode = (genderString: UserPivot['gender']): APIUser['gender'] => {
  if (genderString === undefined || genderString.length === 0) return undefined;
  return genderString === 'male' ? 1 : 2;
};
