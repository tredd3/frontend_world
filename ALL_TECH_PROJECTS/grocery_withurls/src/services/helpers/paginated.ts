const mergeArrays = <T>(previousArray: T[], nextArray: T[]): T[] => (
  [...previousArray, ...nextArray]
);

export type Paginated<T> = {
  data: T | null;
  loadNextPage: null | (() => Promise<Paginated<T>>);
}

// So admittedly, this function looks complicated. But that's only because
// there's a ton of annotations to make it typesafe. The usage is actually
// rather simple. See the tests to see detailed examples.
// This function has two overloads - one that returns arrays, and one that
// returns any other type.
//
// When returning arrays:
//   ```
//   const paginatedService = paginated(
//     (pageNumber = 0, arg1: num, arg2: str) => {
//       return {
//         data: await makeApiCall(/* ... */), // returns Promise<Array<T>>
//         nextPage: pageNumber + 1
//       }
//     }
//   );
//   ```
// Now, `paginatedService`, which will be used in the app, has the
// following signature:
// `(arg1: num, arg2: str) => Promise<{ data: Array<T>, nextPage: null | fn }>`
// That is, you invoke it with the args, and await for data and nextPage.
// `nextPage` is null or a fn. If null, we're done with pagination.
// If fn, we can get more data by calling the fn, which continues to return
// the same type Promise<{ data: Array<T>, nextPage: null | fn }>.
// `data` contains the cumulative data from all pages, so you don't have to
// aggregate it yourself in component state.
//
// When returning non-arrays:
//
//   ```
//   const paginatedService = paginated(
//     (pageNumber = 0, arg1: num, arg2: str) => {
//       return {
//         data: await makeApiCall(/* ... */), // return Promise<any !extends []>
//         nextPage: pageNumber + 1
//       }
//     },
//     (a, b) => { /* merge a and b */}
//   );
//   ```
// The service call itself is similar to the previous example, except
// makeApiCall doesn't return an array in this case. This is the case with
// say getKiranas. In such a situation, you also need to specify how to merge
// previous results with the next results, which you do by providing a second
// `merge` function. The return type of invoking `paginatedService` is the
// same as above, except data will obviously not be an array.
//
// When you don't have any more data to return, simply return null.
//
// In both examples above, the `pageNumber` above is specified by as a number
// but that's not necessary. In fact, it could be of any type, as long as the
// `nextPage` has the same type as `pageNumber`. This is allow you to have
// different pagination parameters per page. (eg. last offset, offset size).

function paginated<P, A extends any[], R>(
  serviceCall: (pageNum: P, ...args: A) =>
    Promise<null | { nextPage: P; data: R[] }>
): (...args: A) => Promise<Paginated<R[]>>
function paginated<P, A extends any[], R>(
  serviceCall: (pageNum: P, ...args: A) =>
    Promise<null | { nextPage: P; data: R }>,
  merge: (a: R, b: R) => R
): (...args: A) => Promise<Paginated<R>>
function paginated <P, A extends any[], R>(
  serviceCall: (pageNum: P, ...args: A) =>
    Promise<null | { nextPage: P; data: R[] | R }>,
  merge?: (a: R, b: R) => R
): (...args: A) => Promise<Paginated<R | R[]>> {
  let pagination: P;
  let data: R | R[] | null = null;

  return (...args: A): Promise<Paginated<R | R[]>> => {
    // Reset internal state incase we're loading this for the first time.
    // This is needed so that if a user goes forward from a page with a listing
    // and comes back, we need to ensure that our internal state is reset.
    if (pagination === null) data = null;

    const loadNextPage = async (): Promise<Paginated<R | R[]>> => {
      const result = await serviceCall(pagination, ...args);

      if (result === null) return { data, loadNextPage: null };
      if (!Array.isArray(result.data) && !merge) {
        throw new Error('You must provide a merge function for non-array return types');
      }

      pagination = result.nextPage;

      if (data === null) { // We don't have any data so far
        data = result.data;
      } else if (Array.isArray(result.data)) { // First overload: array return
        data = mergeArrays((data || []) as R[], result.data);
      } else { // Second overload: non-array return
        data = data ? merge!(data as R, result.data) : result.data;
      }

      return { data, loadNextPage: pagination ? loadNextPage : null };
    };

    return loadNextPage();
  };
}

export default paginated;
