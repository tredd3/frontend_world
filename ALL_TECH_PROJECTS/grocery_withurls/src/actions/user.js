import { SET_USER, NEW_USER, SET_NEW_AUTH_TOKEN } from './constants';
import { checkServiceability } from './address';
import { getUser as getUserService, addUser as addUserService } from '../services/user';
import { getConfig } from '../services/config';
import updateUserAction from './update-user';
import { APIError } from '../services/types';

export const updateUser = updateUserAction;

export const setUser = data => ({
  type: SET_USER,
  data
});

export const newUser = data => ({
  type: NEW_USER,
  data
});

export const setNewToken = data => ({
  type: SET_NEW_AUTH_TOKEN,
  data
});

export const setNewUser = () => newUser({ newUser: true });

export const addUser = ({ firstName, lastName }) => async dispatch => {
  try {
    await addUserService({ firstName, lastName });
    getUserService.clear();
    const user = await getUserService();
    dispatch(setUser(user));
    dispatch(newUser({ newUser: false, newUserErrorMessage: '' }));
    try {
      await getConfig();
    } catch (e) {
      console.log(e.message);
    }
  } catch (e) {
    if (e instanceof APIError) {
      dispatch(newUser({ newUserErrorMessage: e.message }));
    }
  }
};

export const getUser = () => async dispatch => {
  try {
    const user = await getUserService();
    return [
      dispatch(setUser(user)),
      dispatch(checkServiceability({ pincode: user.preferences.pincode }))
    ];
  } catch (e) {
    return [
      dispatch(setUser({}))
    ];
  }
};
