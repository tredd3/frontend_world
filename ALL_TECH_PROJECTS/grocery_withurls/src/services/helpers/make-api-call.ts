/* eslint-disable no-console */
import uuidV1 from 'uuid/v1';
import { needsEnvelope, needsAuthentication, v2ApiRoutes } from '../api-routes';
import { getTokens } from '../../intents';
import { HTTPError, APIError } from '../types';

export type ControlType = { Status: number; MessageCode: number; Message: string };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const prepareBody = <RequestBodyType extends Record<string, any>>(body: RequestBodyType, url: string) => {
  const requestBody = {
    Request: {
      Header: {
        DeviceId: '6cf35d17989d3fbe',
        RequestTime: new Date().getTime(),
        RequestId: uuidV1(),
        Version: v2ApiRoutes.includes(url) ? 2.0 : parseFloat(process.env.REACT_APP_API_VERSION!),
        Product: Number(process.env.REACT_APP_PRODUCT_VERSION)
      },
      Body: body || {}
    }
  };

  return needsEnvelope(url) ? { Data: requestBody } : requestBody;
};

type LogType = 'success' | 'error';
type LogInfo = {
  startTime: number;
  url: string;
  fetchOptions: any;
  response: any;
  json?: any;
  type: LogType;
};
const log = ({
  startTime, url, fetchOptions, response, json, type
}: LogInfo) => {
  if (process.env.NODE_ENV === 'development') { // Compiles this block out in non-dev builds
    const { body, ...otherOptions } = fetchOptions;
    console.groupCollapsed(
      `HTTP request: ${url} %c${Date.now() - startTime}ms %c${type}`,
      'color: grey;',
      `color: ${type === 'success' ? 'green' : 'red'}`
    );
    console.log('Request body:', JSON.stringify(JSON.parse(body), null, 2));
    if (json) console.log('Response json', JSON.stringify(json, null, 2));
    console.log('Full request:', { body, ...otherOptions });
    console.log('Full response:', response);
    console.groupEnd();
  }
};

const doesBodyContainError = ({ Control }: { Control: ControlType }) => (
  !Control || Control.Status === 0 || Control.MessageCode === 404
);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MakeApiCallOptions<RequestBodyType extends Record<string, any>> =
  Partial<Omit<RequestInit, 'body'>> & { body: RequestBodyType };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default async <RequestBodyType extends Record<string, any>, ResponseType>(
  url: string,
  { headers, body, ...otherOptions }: MakeApiCallOptions<RequestBodyType> = { body: {} as RequestBodyType }
) => {
  const tokens = await getTokens();

  const startTime = Date.now();
  const fetchOptions = {
    method: 'POST',
    headers: {
      'X-API-Key': tokens.apiKey,
      'Content-Type': 'application/json',
      'X-REQUEST-ID': uuidV1(),
      'app-name': 'RJIL_JioKart',
      ...(needsAuthentication(url) && { Authorization: `Bearer ${tokens.accessToken}` }),
      ...headers
    },
    body: JSON.stringify(prepareBody<RequestBodyType>(body, url)),
    ...otherOptions
  };
  const response = await fetch(url, fetchOptions);

  if (!response.ok) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, type: 'error'
      });
    }
    throw new HTTPError(response);
  }
  const json = await response.json();

  if (doesBodyContainError(json)) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, json, type: 'error'
      });
    }
    throw new APIError(json);
  }

  if (process.env.NODE_ENV === 'development') {
    log({
      startTime, url, fetchOptions, response, json, type: 'success'
    });
  }

  return json.Data ? json.Data as ResponseType : json.Control as ResponseType;
};
