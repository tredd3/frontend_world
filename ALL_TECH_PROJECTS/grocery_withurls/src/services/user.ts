import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { formatDate } from '../helpers/functional';
import { pivotAPIUser } from './helpers/pivots';
import { APIUser, UserPivot } from '../types';
import { genderStringToCode } from './helpers/generic';

export const getUser = promiseMemoize(async (): Promise<UserPivot> => {
  const user = await makeApiCall<{}, APIUser>(apiRoutes.getUserDetail);
  return pivotAPIUser(user);
});

type UpdateUserArgs = {
  userId?: number;
  firstName?: string;
  lastName?: string;
  email?: string;
  gender?: UserPivot['gender'];
  dob?: Date;
  lat?: number;
  lng?: number;
  pincode?: number;
  city?: string;
  storeId?: number;
  storeName?: string;
};

export const updateUser = async ({
  firstName, lastName, email, gender, dob, lat, lng, pincode, city, storeId, storeName
}: UpdateUserArgs) => {
  const userFromApi = await getUser();
  const { userId } = userFromApi;
  const isValidStoreId = storeId !== undefined;
  const isValidStoreName = storeName !== undefined;

  await makeApiCall(apiRoutes.updateUser, {
    body: {
      userId,
      ...(firstName && { firstName }),
      ...(lastName && { lastName }),
      ...(email ? { emailId: email } : { emailId: '' }),
      ...(gender && { gender: genderStringToCode(gender) }),
      ...(dob ? { dob: formatDate(dob) } : { dob: '' }),
      preferences: {
        ...userFromApi.preferences,
        ...(pincode && { pincode }),
        ...(lat && { latitude: lat }),
        ...(lng && { longitude: lng }),
        ...(isValidStoreId && { storeId }),
        ...(city && { city }),
        ...(isValidStoreName && { storeName })
      }
    }
  });

  getUser.clear();
};

type AddUserArgs = {
  firstName: string;
  lastName: string;
};

export const addUser = ({ firstName, lastName }: AddUserArgs) => (
  makeApiCall<AddUserArgs, void>(apiRoutes.addUser, {
    body: {
      firstName,
      lastName
    }
  })
);
