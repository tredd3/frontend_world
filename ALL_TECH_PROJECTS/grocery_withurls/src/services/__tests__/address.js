import mockDB from '../../mocks/db.json';
import * as UserService from '../user';
import * as makeApiCall from '../helpers/make-api-call';
import { apiRoutes } from '../api-routes';
import {
  getAddresses, addAddress, isAreaServiceable, updateAddress, deleteAddress, getDefaultAddress, getAddressById,
  updateAddressWithKirana, deleteKiranaFromAddress
} from '../address';

const { getUser } = UserService;

jest.mock('../../intents', () => ({
  getTokens: () => ({
    apiKey: 'dummy-api-key',
    accessToken: 'dummy-access-token'
  })
}));

describe('address service mocked tests', () => {
  let makeApiCallSpy;
  let getUserSpy;

  beforeAll(() => {
    getUserSpy = jest.spyOn(UserService, 'getUser').mockImplementation(async () => mockDB.user.Data);
    makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.saveUserAddress.Data);
  });

  beforeEach(fetch.resetMocks);

  it('addAddress should prepare `address` string field and add it while making API request', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.saveUserAddress));

    const mockedAddress = mockDB.saveUserAddress.Data;
    const addressPayload = { ...mockedAddress };
    delete addressPayload.address;

    await addAddress({
      ...addressPayload,
      lat: mockDB.saveUserAddress.Data.latitude,
      lng: mockDB.saveUserAddress.Data.longitude
    });

    expect(makeApiCallSpy).toHaveBeenLastCalledWith(
      apiRoutes.saveAddress,
      { body: expect.objectContaining({ address: `${mockedAddress.cityName}, ${mockedAddress.pincode}` }) }
    );
  });

  it('updateAddress should prepare `address` string field and add it while making API request', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.saveUserAddress));

    const mockedAddress = mockDB.saveUserAddress.Data;
    const addressPayload = { ...mockedAddress };
    delete addressPayload.address;

    await updateAddress({
      ...addressPayload,
      lat: mockDB.saveUserAddress.Data.latitude,
      lng: mockDB.saveUserAddress.Data.longitude,
      id: '123'
    });

    expect(makeApiCallSpy).toHaveBeenLastCalledWith(
      apiRoutes.saveAddress,
      { body: expect.objectContaining({ address: `${mockedAddress.cityName}, ${mockedAddress.pincode}` }) }
    );
  });

  afterAll(() => {
    makeApiCallSpy.mockRestore();
    getUserSpy.mockRestore();
  });
});

describe('update address with kirana and delete kirana from address tests', () => {
  let updateUserSpy;

  beforeAll(() => {
    updateUserSpy = jest.spyOn(UserService, 'updateUser').mockImplementation(async () => mockDB.user.Data);
  });

  beforeEach(() => {
    getUser.clear();
    getAddresses.clear();
    fetch.resetMocks();
    updateUserSpy.mockClear();
  });

  it('should update user on adding kirana to default address', async () => {
    const mockAddresses = JSON.parse(JSON.stringify(mockDB.address));
    mockAddresses.Data.addresses[0].isDefault = true;

    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockAddresses));
    fetch.mockResponseOnce(JSON.stringify({
      Control: { Status: 1 },
      Data: mockAddresses.Data.addresses[0]
    }));

    await updateAddressWithKirana(mockAddresses.Data.addresses[0].addressId, { storeId: 1234, storeName: 'test store' });
    expect(updateUserSpy).toHaveBeenLastCalledWith({
      storeId: 1234,
      storeName: 'test store'
    });
  });

  it('should skip updating user on adding kirana to non-default address', async () => {
    const mockAddresses = JSON.parse(JSON.stringify(mockDB.address));
    mockAddresses.Data.addresses[0].isDefault = false;

    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockAddresses));
    fetch.mockResponseOnce(JSON.stringify({
      Control: { Status: 1 },
      Data: mockAddresses.Data.addresses[0]
    }));

    await updateAddressWithKirana(mockAddresses.Data.addresses[0].addressId, { storeId: 1234, storeName: 'test store' });
    expect(updateUserSpy).not.toHaveBeenCalled();
  });

  it('should update user on deleting kirana from default address', async () => {
    const mockAddresses = JSON.parse(JSON.stringify(mockDB.address));
    mockAddresses.Data.addresses[0].isDefault = true;

    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockAddresses));
    fetch.mockResponseOnce(JSON.stringify({
      Control: { Status: 1 },
      Data: mockAddresses.Data.addresses[0]
    }));

    await deleteKiranaFromAddress(mockAddresses.Data.addresses[0].addressId);
    expect(updateUserSpy).toHaveBeenLastCalledWith({
      storeId: 0,
      storeName: ''
    });
  });

  it('should skip updating user on deleting kirana from address which is not default', async () => {
    const mockAddresses = JSON.parse(JSON.stringify(mockDB.address));
    mockAddresses.Data.addresses[0].isDefault = false;

    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockAddresses));
    fetch.mockResponseOnce(JSON.stringify({
      Control: { Status: 1 },
      Data: mockAddresses.Data.addresses[0]
    }));

    await deleteKiranaFromAddress(mockAddresses.Data.addresses[0].addressId);
    expect(updateUserSpy).not.toHaveBeenCalled();
  });

  afterAll(() => {
    updateUserSpy.mockRestore();
  });
});

describe('address service', () => {
  beforeEach(fetch.resetMocks);
  afterEach(getUser.clear);

  it('checkServiceablity should return true when we get a valid response', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.isPincodeServiceable));
    isAreaServiceable.clear();

    const isServiceable = await isAreaServiceable({ pincode: 400701 });
    expect(isServiceable).toBe(true);
  });

  it('checkServiceablity should return false when we get an error response', async () => {
    fetch.mockResponseOnce(JSON.stringify({
      Control: {
        Message: 'Currently service not available in your pincode 560103. Try a different location',
        MessageCode: 200,
        Status: 0
      }
    }));
    isAreaServiceable.clear();

    const isServiceable = await isAreaServiceable({ pincode: 400412 });
    expect(isServiceable).toBe(false);
  });

  it('getAddresses should give back correct response', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.address));
    const addresses = await getAddresses();
    expect(addresses).toMatchSnapshot();
  });

  it('getDefaultAddress should return the default response', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.address));
    const defaultAddress = await getDefaultAddress();
    expect(defaultAddress.id).toBe(10160497);
  });

  it('getAddressById should return the correct address', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.address));
    const address = await getAddressById(mockDB.address.Data.addresses[0].addressId);
    expect(address.id).toBe(mockDB.address.Data.addresses[0].addressId);
  });

  it('addAddress should give back correct response', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.saveUserAddress));
    const newAddress = await addAddress({
      ...mockDB.saveUserAddress.Data,
      lat: mockDB.saveUserAddress.Data.latitude,
      lng: mockDB.saveUserAddress.Data.longitude
    });
    expect(newAddress).toMatchSnapshot();
  });

  it('updateAddress should give back the correct response', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.saveUserAddress));
    const updatedAddress = await updateAddress({
      ...mockDB.saveUserAddress.Data,
      lat: mockDB.saveUserAddress.Data.latitude,
      lng: mockDB.saveUserAddress.Data.longitude,
      id: '123'
    });
    expect(updatedAddress).toMatchSnapshot();
  });

  it('deleteAddress should give back the correct response', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.removeUserAddress));
    await deleteAddress('123');
    expect(JSON.parse(fetch.mock.calls[0][1].body).Request.Body).toHaveProperty('addressId');
  });
});
