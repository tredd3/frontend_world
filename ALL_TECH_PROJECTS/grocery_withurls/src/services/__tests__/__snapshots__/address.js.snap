// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`address service addAddress should give back correct response 1`] = `
Object {
  "active": 1,
  "addedDate": 2020-01-04T06:24:00.000Z,
  "address": "Navi Mumbai, 400710",
  "addressLine1": "Ramada",
  "addressLine2": "Mahape Road",
  "addressLine3": "Kopar Khairane",
  "addressTag": "Home",
  "cityName": "Navi Mumbai",
  "customerMobileNumber": 9100100100,
  "defaultStore": 4055192,
  "defaultStoreName": "Mamata Kirana And General Stores",
  "firstName": "Firstname",
  "id": 10001886,
  "isAutoSelected": true,
  "isDefault": true,
  "isServiceable": true,
  "landmark": "",
  "lastName": "Lastname",
  "lastUpdatedDate": "2020-01-06 13:56 PM IST",
  "latitude": 19.11224091559444,
  "longitude": 19.11224091559444,
  "phoneNumber": 9100100100,
  "pincode": 400710,
  "storeId": 1040356,
  "storeName": "Prime Medico",
  "userId": 10005564,
}
`;

exports[`address service getAddresses should give back correct response 1`] = `
Array [
  Object {
    "active": 1,
    "addedDate": 2020-02-20T07:56:00.000Z,
    "address": "123 ered",
    "addressLine1": "123",
    "addressLine2": "ered",
    "addressLine3": "Kharghar, Sector 29",
    "addressTag": "Home",
    "addressType": 1,
    "cityName": "Navi Mumbai",
    "customerMobileNumber": 9999999999,
    "defaultStore": 4001313,
    "defaultStoreName": "Bulbasuar Stores",
    "firstName": "Firstname",
    "id": 10160497,
    "isDefault": true,
    "isServiceable": true,
    "lastName": "lastname",
    "lastUpdatedDate": "2020-02-26 15:02 PM IST",
    "latitude": 19.0630231,
    "longitude": 73.0700421,
    "phoneNumber": 9999999999,
    "pincode": 410210,
    "storeId": 7001290,
    "storeName": "Vedavati Stores",
    "userId": 10041541,
  },
  Object {
    "active": 1,
    "addedDate": 2020-02-26T10:37:00.000Z,
    "address": "s s",
    "addressLine1": "sStry a long long name for address",
    "addressLine2": "s",
    "addressLine3": "Kharghar, Sector 29",
    "addressTag": "home",
    "cityName": "Navi Mumbai",
    "customerMobileNumber": 9999999999,
    "defaultStore": 4001313,
    "defaultStoreName": "Bulbasuar Stores",
    "firstName": "Firstname",
    "id": 10160580,
    "isDefault": false,
    "isServiceable": true,
    "landmark": "s",
    "lastName": "lastname",
    "lastUpdatedDate": "2020-02-26 17:48 PM IST",
    "latitude": 19.0630231,
    "longitude": 73.0700421,
    "phoneNumber": 9999999999,
    "pincode": 410210,
    "storeId": 0,
    "userId": 10041541,
  },
]
`;

exports[`address service updateAddress should give back the correct response 1`] = `
Object {
  "active": 1,
  "addedDate": 2020-01-04T06:24:00.000Z,
  "address": "Navi Mumbai, 400710",
  "addressLine1": "Ramada",
  "addressLine2": "Mahape Road",
  "addressLine3": "Kopar Khairane",
  "addressTag": "Home",
  "cityName": "Navi Mumbai",
  "customerMobileNumber": 9100100100,
  "defaultStore": 4055192,
  "defaultStoreName": "Mamata Kirana And General Stores",
  "firstName": "Firstname",
  "id": 10001886,
  "isAutoSelected": true,
  "isDefault": true,
  "isServiceable": true,
  "landmark": "",
  "lastName": "Lastname",
  "lastUpdatedDate": "2020-01-06 13:56 PM IST",
  "latitude": 19.11224091559444,
  "longitude": 19.11224091559444,
  "phoneNumber": 9100100100,
  "pincode": 400710,
  "storeId": 1040356,
  "storeName": "Prime Medico",
  "userId": 10005564,
}
`;
